wordpress-mobile/WordPress-Android
# Add media upload completion processor

Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/1629

### Description

This PR is the Android implementation of a new processor which can handle media upload completion events while the Gutenberg editor is closed.

This uses a "hybrid approach", with both RegEx, and the jsoup html parser to parse and splice the post content and update local media ids and urls for various cases. This addresses some of the limitations of the previous approach.

This approach borrows from the previous approach, and also the approach used for iOS: https://github.com/wordpress-mobile/WordPress-iOS/pull/13032.

#### Possible improvements for future iterations:

The implementation in this PR can be improved (and DRYed up) by using a JSON parser to handle the block header contents in a more general way.

#### Note:

The HTML parser used in this implementation has two modes with slightly different serialization outputs. However, post content I've encountered in my tests conforms strictly to neither:

Content source|self-closing-tags|boolean-attributes
-|-|-
`html` mode|`<img>`|`checked`
`xml` mode|`<img />`|`checked=""`
Post content|`<img />`|`checked`

I don't foresee this being a problem, but it's worth keeping in mind while writing any new tests.

### Tests

Parts of this implementation make extensive use of RegEx patterns, which can be difficult to read. I've added javadocs in many areas to mitigate that. Also, I've added a suite of unit tests to cover many scenarios.

#### Copied from the @SergioEstevao 's iOS PR:

It handles 3 possible scenarios for galleries, that depend of the link to value:
 - none: there is no `a` element is in the gallery and we just have img
 - media:  we have a `a` element and it points to the full url of the img
 - attachment: we have a 'a' element and it should point to the attachment page for media. I noticed that the moment the gallery block is not inserting the `a` element because the `link` field for media is not defined. Fixing this should be done on the main gallery block PR because we will need to start to send the link information for image/media to the JS/React side. This will need to change to the bridge code and in the main apps to retrieve that information.

#### To test:
 - Create a new post using Gutenberg
 - Add a gallery block with some images
 - While the images are still uploading press publish
 - Wait outside the editor for the post to be published
 - Open the post again
 - Check that the gallery block is working
 - Switch to HTML mode and see if the link and id are pointing to the server/remote values.

Also, please test this same flow for image and media-text blocks.

#### Video block

This PR may address some of the issues described here: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1512, but this needs further testing.

PR submission checklist:
